### ISSUE TITLE
Highlight or Prevent Heading Level Errors in Markdown 

### ISSUE THREAD

**Issue by aitchiss** [forem/forem#4807]
**Is your feature request related to a problem? Please describe.**

This issue is coming out of a discussion on post https://dev.to/s_aitchison/getting-heading-levels-right-including-here-on-dev-439

A substantial proportion of posts on DEV break WCAG guidance on heading levels. The post title automatically becomes the `h1` of the page; from that point on heading levels should only increase by 1. 

What you actually see in reality is a lot of additional `h1` elements, and then quite often non-sequential heading levels beyond that.

It would be a boost for accessibility if we could help limit or prevent this in some way. As a bit of context, the latest WebAIM screen reader user survey indicated that close to 70% of users prefer to navigate by heading level, and the ability to scan content this way is compromised by non-WCAG compliant heading usage.

**Describe the solution you'd like**

When viewing a post in 'preview' mode, it would be good to see heading level issues flagged in the "Heads up" warning that appears at the top of the page. At a minimum, flagging duplicate `h1`s, but ideally linting for other issues in the heading levels (e.g. an `h3` being present but no `h2`)

**Comment by colabottles** [comment]
@nickytonline I can take a look at this if you want to let me have a crack at this one.

**Comment by techieeliot** [comment]
Hey @nickytonline, I'd like to tackle this one tomorrow.

**Comment by nickytonline** [comment]
Sounds good. I've assigned this to you.

**Comment by techieeliot** [comment]
# Does this look like an H1 in github

Note that we're looking at options: 

- markdown-lint  - https://github.com/DavidAnson/markdownlint

**Comment by techieeliot** [comment]
## What we discovered on the live stream

- @nickytonline @cmgorton and @techieeliot we toyed around with the markdown lint demo from the docs
- in the docs we tested out the following code in the articleForm.js at `app/javascript/packs/`: 

```
import markdownlint from 'markdown-lint';

const options = {
  "strings": {
    "good.string": "# good.string\n\nThis string passes all rules.",
    "bad.string": "#bad.string\n\n#This string fails\tsome rules."
  }
};

markdownlint(options, function callback(err, result) {
  if (!err) {
    console.log(result.toString());
  }
});
```

-  we discovered after trial and error that we did not need to import it, so we removed it and swiped the following docs from the source and placed them into the path `app/assets/javascripts` then applied the following into the base.js.erb: 

```
//= require markdown-it.min.js
//= require markdownlint-browser.min.js
```

- Once we applied the minified markdown and markdownlint we were able to see that the markdownlint was accessible inside the browser now without import and the function was calling to show that the good and bad scripts.

**Comment by techieeliot** [comment]
I went to commit on my branch, and I had these errors in my terminal:
```
husky > pre-commit (node v12.18.4)
ting to original state because of errors...
✔ Cleaning up... 

✖ eslint --fix:

/Users/techieeliot/Desktop/projects/forem/app/assets/javascripts/markdown-it.min.js
     5:5   error  'module' is not defined                                                    no-undef
     6:43  error  'define' is not defined                                                    no-undef
     6:55  error  'define' is not defined                                                    no-undef
    11:9   error  'global' is not defined                                                    no-undef
    22:51  error  'require' is not defined                                                   no-undef
    45:47  error  'require' is not defined                                                   no-undef
   177:39  error  Unnecessary escape character: \/                                           no-useless-escape
   488:27  error  Empty block statement                                                      no-empty
   496:27  error  Empty block statement                                                      no-empty
  2539:22  error  Expected a conditional expression and instead saw an assignment            no-cond-assign
  2611:42  error  Unnecessary escape character: \/                                           no-useless-escape
  2612:17  error  Unexpected control character(s) in regular expression: \x00                no-control-regex
  2612:41  error  Unnecessary escape character: \-                                           no-useless-escape
  2725:21  error  Do not access Object.prototype method 'hasOwnProperty' from target object  no-prototype-builtins
  3223:18  error  Expected a conditional expression and instead saw an assignment            no-cond-assign
  5766:31  error  Do not access Object.prototype method 'hasOwnProperty' from target object  no-prototype-builtins
  6200:33  error  Unnecessary escape character: \?                                           no-useless-escape
  6249:43  error  Unnecessary escape character: \/                                           no-useless-escape
  6249:51  error  Unnecessary escape character: \/                                           no-useless-escape
  6515:37  error  Do not access Object.prototype method 'hasOwnProperty' from target object  no-prototype-builtins
  6521:17  error  'global' is not defined                                                    no-undef
  6533:23  error  Unexpected control character(s) in regular expression: \x1f                no-control-regex
  6545:30  error  Unnecessary escape character: \*                                           no-useless-escape
  6545:34  error  Unnecessary escape character: \/                                           no-useless-escape
  6545:38  error  Unnecessary escape character: \?                                           no-useless-escape
  6545:41  error  Unnecessary escape character: \[                                           no-useless-escape
  6545:47  error  Unnecessary escape character: \{                                           no-useless-escape
  6545:49  error  Unnecessary escape character: \}                                           no-useless-escape

/Users/techieeliot/Desktop/projects/forem/app/assets/javascripts/markdownlint-browser.min.js
     4:8   error  'module' is not defined                                          no-undef
     5:38  error  'define' is not defined                                          no-undef
     6:7   error  'define' is not defined                                          no-undef
    10:11  error  'global' is not defined                                          no-undef
    20:51  error  'require' is not defined                                         no-undef
    43:47  error  'require' is not defined                                         no-undef
   442:17  error  't' is already defined                                           no-redeclare
   517:47  error  'n' is already defined                                           no-redeclare
   627:29  error  Empty block statement                                            no-empty
  1365:17  error  'e' is already defined                                           no-redeclare
  1619:25  error  Expected a conditional expression and instead saw an assignment  no-cond-assign
  1799:17  error  'a' is already defined                                           no-redeclare
  1801:17  error  'u' is already defined                                           no-redeclare
  1813:33  error  'o' is already defined                                           no-redeclare
  1929:17  error  'n' is already defined                                           no-redeclare
  2278:32  error  Expected a conditional expression and instead saw an assignment  no-cond-assign
  2328:17  error  'n' is already defined                                           no-redeclare
  2684:21  error  Unexpected constant condition                                    no-constant-condition
  2716:17  error  'x' is already defined                                           no-redeclare
  3221:34  error  'l' is already defined                                           no-redeclare
  3480:21  error  't' is already defined                                           no-redeclare
  3518:42  error  Unexpected console statement                                     no-console
  3643:45  error  Unexpected control character(s) in regular expression: \x1b      no-control-regex
  3838:15  error  Unexpected console statement                                     no-console
  3861:17  error  'global' is not defined                                          no-undef
  3878:10  error  'module' is not defined                                          no-undef
  3879:40  error  'define' is not defined                                          no-undef
  3880:9   error  'define' is not defined                                          no-undef
  3884:13  error  'global' is not defined                                          no-undef
  3894:53  error  'require' is not defined                                         no-undef
  3917:49  error  'require' is not defined                                         no-undef
  4306:19  error  't' is already defined                                           no-redeclare

/Users/techieeliot/Desktop/projects/forem/app/javascript/packs/articleForm.jsx
  13:1  error  'markdownlint' is not defined  no-undef
  15:5  error  Unexpected console statement   no-console

✖ 62 problems (62 errors, 0 warnings)```

**Comment by rhymes** [comment]
Hi @techieeliot, I defer to @nickytonline on the implementation, just wanted to add that if that library is an external dependency it should be installed as such, not added to the source code directly. Basically, something like `yarn add -D markdownlint`

**Comment by nickytonline** [comment]
@rhymes, there is an npm package, which works in Node, but to use it in the frontend, you need to build it and load the scripts the old school way. On the stream, we popped it in base.js.erb just to get the POC to see it working.

The actual implementation will most likely load them dynamically once the editor is loaded so as to prevent page slowdowns.

**Comment by nickytonline** [comment]
@techieeliot, for the eslint errors, I can take a peek.

**Comment by nickytonline** [comment]
Ahh, so for the eslint errors, we just need to exlcude these files from eslint as it is not code we manage even though we're including it in the project. Just add those two files to https://github.com/forem/forem/blob/master/.eslintignore

**Comment by techieeliot** [comment]
Ok, thanks @nickytonline! Will do.

**Comment by techieeliot** [comment]
Hey, I'm going get back on this one. I'll aim to have something to up on this one.

**Comment by cmgorton** [comment]
@nickytonline maybe a potential RFC to encourage @aitchiss to propose since they opened this issue?

**Comment by aitchiss** [comment]
This has been closed in #13745 🎉  Thanks @techieeliot for laying the groundwork on this one!
